types/artifacts.d.ts

```diff
@@ -12,177 +14,177 @@ import speedline from 'speedline-core';
 import TextSourceMap from '../core/lib/cdt/generated/SourceMap.js';
 import {ArbitraryEqualityMap} from '../core/lib/arbitrary-equality-map.js';
 import type { TaskNode as _TaskNode } from '../core/lib/tracehouse/main-thread-tasks.js';
-import AuditDetails from './lhr/audit-details'
-import Config from './config';
-import Gatherer from './gatherer';
-import {IcuMessage} from './lhr/i18n';
-import LHResult from './lhr/lhr'
-import Protocol from './protocol';
+import AuditDetails from './lhr/audit-details.js'
```

Comment:
Are the `.js` extensions from an earlier iteration of the PR? I don't believe they're necessary and not having an extension is a nice signal that it's importing a non-source file (vs tsc's magic handling of `.js` paths leading to `.ts` files)

Reply:
In my testing they were necessary once the package is installed as a dependency. Not 100% sure why though.

Reply:
So I did a little further investigation. Importing types from `'./utility-types'` without the `.js` extension works fine, however importing types from `'./user-flow'` without the `.js` extension causes issues.

I think it's related to `user-flow.d.ts` importing types from an actual `.js` file.

Reply:
What error are you seeing? I removed all the `.js`s from imports where there wasn't actually a `.js` file and the only errors I get when importing lighthouse (trying both default `lighthouse()` and `startFlow()`) are on global `LH` and puppeteer conflicts

Reply:
When importing the package into another project, we don't get any type checking info.

Basically any type imported from a file not using the .js extension is treated as `any`

Reply:
<img width="763" alt="example lighthouse code showing a typescript error on a function parameter" src="https://user-images.githubusercontent.com/316891/206786561-c7d209d7-7398-4cfa-ac34-af6f2d5dbf54.png">

working for me, and this goes
`import {startFlow} from 'lighthouse'`
`index.d.ts` -> `import * as LH from "../types/lh.js"` to get `LH.UserFlow.Options`
`lh.d.ts` -> `import UserFlow_ from './user-flow'` to get `UserFlow`
`user-flow.d.ts` -> `import {Flags} from './externs'` to get the `Flags` type

so two hops through imports with no file extension, and `user-flow.d.ts` has imports with and without an extension, because it needs the actual js `'../core/user-flow.js'` file.

Reply:
I think I figured out what's going on...

I was doing my testing in a project with a `tsconfig.json` containing `"moduleResolution": "NodeNext"`. If I change it to `"moduleResolution": "node"` my results match what you are seeing.

Seems like we need the `.js` extension on our module imports so type checking works in ts projects set with `"moduleResolution": "NodeNext"`.

Reply:
> Seems like we need the .js extension on our module imports so type checking works in ts projects set with "moduleResolution": "NodeNext".

That's the purpose of nodenext - to opt into standards compliant ES modules. No need to guess the extension, and TS prefers to not modify authored code when possible so they chose to require .js and skip any rewriting step based on the output format.

Reply:
@brendankenny WDYT about keeping the `.js` extensions for compatibility with `NodeNext`?

Reply:
Seems good from a compatibility perspective to keep the `.js` extensions.

Reply:
Ugh, I guess there's not any way around this.

Just to be clear about my objection:
- Unlike the usual `.ts` -> `.js` import, there is no js file to import, and there won't be after a build step, because these are imports of `d.ts` files purely for types. Even if we were doing a pure-ts compile, `d.ts` files never get compiled or emitted.
- there's no issue of the compiler not re-writing anything because these imports are _also_ in `d.ts` files, which again are never compiled or emitted, so there's no output to rewrite the paths to

As far as `tsc` is concerned, these files only exist at type-checking time and then cease to exist, and they are definitely not es modules.

`nodenext` still has a ton of issues (a number of fixes queued up before 5.0), but as far as I can tell they aren't going to budge on this unless a more significant contingent of jsdoc users object. I guess we'll have to live with these js imports.